feat: add C implementation for `@stdlib/stats/base/dists/lognormal/logcdf`

saurabhraghuvanshii opened the pull request:
Resolves #3750 .

## Description

> What is the purpose of this pull request?

This pull request:

-   Add C implementation for `@stdlib/stats/base/dists/lognormal/logcdf`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3750 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

test case failed , I try multiple things.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

saurabhraghuvanshii committed:
feat: add C implementation for base/dists/lognormal/logcdf

saurabhraghuvanshii committed:
chore: update copyright year

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

saurabhraghuvanshii commented:
some errors are showing in readme section, And I don't know to tackle with that, I need help and review and Tests are failed  as I done every type of possible solution by my side .

saurabhraghuvanshii commented:
@Planeshifter @kgryte  review this , I have some doubt with this.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/src/main.c:
```diff
@@ -0,0 +1,47 @@
+* double y = stdlib_base_lognormal_logcdf( 2.0, 0.0, 1.0 );
+* // returns ~-0.2799
+*/
+double stdlib_base_dists_lognormal_logcdf( const double x, const double mu, const double sigma ) {
```
@saurabhraghuvanshii The test cases are throwing errors because your implementation for `main.c` file does not resonate with the `lib/main.js`
According to the `lib/main.js` you'll be needing C implementations for `nlogcdf`

saurabhraghuvanshii replied:
means ` return NINF / ln(x) `, if yes then how import ln it show there is no directory

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/src/main.c:
```diff
@@ -0,0 +1,47 @@
+		stdlib_base_is_nan( sigma ) ||
+		sigma <= 0.0
+	) {
+		return 0.0 / 0.0; // NaN
```
it should be returning `NINF` when true and 
```c
ln(x)
 ```
 while false

saurabhraghuvanshii replied:
`#include "stdlib/stats/base/dists/lognormal/logcdf.h"
#include "stdlib/math/base/special/ln.h"
#include "stdlib/constants/float64/ninf.h"

/**
* Returns the logcdf for a lognormal distribution with mean `mu` and standard deviation `sigma`.
*
* @param x         input value
* @param mu        mean
* @param sigma     standard deviation
* @return          logcdf
*
* @example
* double y = stdlib_base_lognormal_logcdf( 2.0, 0.0, 1.0 );
* // returns ~-0.2799
*/
double stdlib_base_dists_lognormal_logcdf( const double x, const double mu, const double sigma ) {
	double lx = ( x <= 0.0 ) ? ninf : ln( x );
	return nlogcdf( lx, mu, sigma );
}
`
error on building

saurabhraghuvanshii replied:
`fatal error: stdlib/math/base/special/ln.h: No such file or directory`

Neerajpathak07 replied:
did you add those header files in `manifest.json` also that's not how you write ninf it should be `STDLIB_CONSTANT_FLOAT64_NINF`

Neerajpathak07 replied:
Similarly instead of `ln(x)` write `stdlib_base_ln(x)`
same goes for nlogcdf

saurabhraghuvanshii replied:
same error happen

saurabhraghuvanshii replied:
can you go through files annd check why all this happens

saurabhraghuvanshii replied:
I am comiting with new changes check it .

Neerajpathak07 replied:
sure

saurabhraghuvanshii replied:
now you can check,

saurabhraghuvanshii committed:
chore: update main.c addon.c manifest.json

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

saurabhraghuvanshii commented:
I have to clean up , I don't how all these commit comes from

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/manifest.json:
```diff
@@ -0,0 +1,83 @@
+        "dependencies": [
+          "@stdlib/math/base/napi/binary",
+          "@stdlib/math/base/assert/is-nan",
+          "@stdlib/math/base/special/exp"
```
remove `is-nan` and `exp` and add the header files for `ln` and `NINF`

Neerajpathak07 replied:
```suggestion
          "@stdlib/constants/float64/ninf"
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/manifest.json:
```diff
@@ -0,0 +1,83 @@
+          "@stdlib/math/base/special/exp",
+          "@stdlib/math/base/special/asin",
+          "@stdlib/math/base/special/sqrt",
+          "@stdlib/constants/float64/pi"
```
you'll have to remove all of these as it has not been used anywhere instead add the files suggested above.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/src/main.c:
```diff
@@ -0,0 +1,38 @@
+#include "stdlib/stats/base/dists/lognormal/logcdf.h"
+#include "stdlib/math/base/special/ln.h"
+#include "stdlib/constants/float64/ninf.h"
+
```
need to add `#include "stdlib/stats/base/dists/normal/logcdf.h"` in order to reference `stdlib_base_dists_normal_logcdf`
But `stdlib/stats/base/dists/normal/logcdf` does not have C implementation so this PR will have to wait till we have that.
You can go ahead and draft this PR for the time being

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/src/main.c:
```diff
@@ -0,0 +1,38 @@
+*/
+double stdlib_base_dists_lognormal_logcdf( const double x, const double mu, const double sigma ) {
+	double lx = ( x <= 0.0 ) ? STDLIB_CONSTANT_FLOAT64_NINF : stdlib_base_ln( x );
+	return stdlib_base_nlogcdf( lx, mu, sigma );
```
This reference is incorrect it should be `stdlib_base_dists_normal_logcdf` as it comes from `stats/base/dists/normal/logcdf`

saurabhraghuvanshii replied:
ok

saurabhraghuvanshii replied:
but we implement for stats/base/dists/lognormal/logcdf

Neerajpathak07 replied:
I'm sorry I didn't understand your point

saurabhraghuvanshii replied:
sorry, I take it wrong, for now I am leaving this pr . in future see

Neerajpathak07 replied:
Though I would suggest you to draft this PR for now

saurabhraghuvanshii replied:
today I try once more and change in manifest from binary to ternary so old error no showing new error are ` code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/saurabh/Gsoc/stdlib-gs/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js',
    '/home/saurabh/Gsoc/stdlib-gs/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js',
    '/home/saurabh/Gsoc/stdlib-gs/lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/[eval]'
  ]
}`

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/ln/manifest.json:
```diff
@@ -38,12 +38,9 @@
       "libraries": [],
       "libpath": [],
       "dependencies": [
-        "@stdlib/math/base/napi/unary",
```
There is no need to make these changes in the `manifest.json` of  `math/base/special/ln`,

saurabhraghuvanshii replied:
there is nothing show like this in `manifest.json` of `math/base/special/ln` in repo

Neerajpathak07 replied:
What I meant was that there was no need to change the block of code in `math/base/special/ln` so it will be good to revert it back.

saurabhraghuvanshii replied:
It is depends on `base/dists/normal/logcdf` , which is not implement at that moment.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/logcdf/manifest.json:
```diff
@@ -0,0 +1,83 @@
+        "libpath": [],
+        "dependencies": [
+          "@stdlib/math/base/napi/binary",
+          "@stdlib/math/base/assert/is-nan",
```
```suggestion
          "@stdlib/math/base/special/ln",
```
This is what I meant by adding the header file for `manifest.json`

saurabhraghuvanshii commented:
closing, outdated and merge conflicts
